### PR TITLE
InstanceMutator LXDProfile backwards compatibility

### PR DESCRIPTION
**Pull request by SimonRichardson**
## Description of change

The following changes allows us to bypass the LXDProfile resolver in the uniter,
so that any changes to the `instanceCharmProfileData` documents state that the
waiting for the LXDProfile is `Not Required`.

With the feature flag enabled it makes future work around LXDProfile workers
possible because the uniter never blocks.

Additionally we now have tests to cover that the new changes always return
`Not Required`.

## QA steps

```
export JUJU_DEV_FEATURE_FLAGS=instance-mutater
juju bootstrap localhost --build-agent
juju deploy ./testcharms/charm-repo/quantal/lxd-profile
juju upgrade-charm lxd-profile --path ./testcharms/charm-repo/quantal/lxd-profile
```

### CODE REVIEW

File: state/watcher.go [hunk 1]
```diff
@@ -2059,22 +2069,33 @@ func (u *Unit) WatchLXDProfileUpgradeNotifications() (StringsWatcher, error) {
 // usage of the not know status.
 // Events are generated when there are changes to a instance charm profile
 // data document.
+//
+// - Watcher DEPRECATED -
+// * The following watcher is deprecated and shouldn't be used, it watches for
+// * changes in the instanceCharmProfileData collection, but instead of
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply state/watcher.go]: So I broke this into two steps:

 1. Update the current workflow to add `lxdprofile.NotRequiredStatus` and add tests
 1. Doing 1 allows for us to then work out how to do it without relying on old code whilst still be behind a feature flag.

File: state/watcher_profilecharm_test.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+	// The done channel is used by tests to indicate that
+	// the worker has accomplished the scenario and can be stopped.
+	done chan struct{}
+	dead chan struct{}
+}
+
+var _ = gc.Suite(&watcherCharmProfileSuite{})
+
+func (s *watcherCharmProfileSuite) SetUpTest(c *gc.C) {
+	s.SetInitialFeatureFlags(feature.InstanceMutater)
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply state/watcher_profilecharm_test.go]: I'll add more tests tbh

**Comment by manadart** [comment]
While you are in there, there's a file name typo here:
/api/instancemutater/+pacakge+_test.go

**Comment by SimonRichardson** [comment]
@manadart sure nps

**Comment by SimonRichardson** [comment]
!!build!!

File: state/watcher.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply state/watcher.go]: I did both...

File: state/database.go [hunk 3]
```diff
@@ -27,6 +27,11 @@ type SessionCloser func()
 
 func dontCloseAnything() {}
 
+//go:generate mockgen -package mocks -destination mocks/database_mock.go github.com/juju/juju/state Database
```
Comment (manadart) [inline state/database.go]: +1. To be used with a vengeance.

File: state/watcher.go [hunk 4]
```diff
@@ -2081,21 +2093,35 @@ func (u *Unit) WatchLXDProfileUpgradeNotifications() (StringsWatcher, error) {
 type instanceCharmProfileDataWatcher struct {
```
Comment (hmlanigan) [inline state/watcher.go]: Suggest renaming to something more generic now that a collection can be specified and isn't hard coded.  As is the field watched.

File: state/watcher.go [hunk 5]
```diff
@@ -2025,38 +2025,50 @@ func (u *Unit) WatchMeterStatus() NotifyWatcher {
...
+		if m, ok := data.(instanceCharmProfileData); ok {
+			return m.UpgradeCharmProfileComplete, nil
+		}
+		return initial, errors.NotValidf("type")
 	}
-	return newInstanceCharmProfileDataWatcher(m.st, watchDocId, returnStatus, filter), nil
+	return newInstanceCharmProfileDataWatcher(backend, collection, members, initial, filter, extract, transform)
 }
 
 // instanceCharmProfileDataWatcher notifies about any changes to the
```
Comment (hmlanigan) [inline state/watcher.go]: are these comments still valid?  If the watcher has a more generic name, maybe others will be able to use it.

File: state/watcher.go [hunk 6]
```diff
@@ -2106,48 +2132,49 @@ func newInstanceCharmProfileDataWatcher(backend modelBackend, memberId, returnSt
...
-			logger.Debugf("instanceCharmProfileData NOT mgo err not found")
+			logger.Debugf("%s NOT mgo err not found", w.collection)
 			return false, err
 		}
-		logger.Tracef("instanceCharmProfileData for %q: mgo err not found", w.docId)
+		logger.Tracef("%s for %v: mgo err not found", w.collection, w.members)
 		return false, nil
 	}
 
 	// check the field before adding to the machineId
```
Comment (hmlanigan) [inline state/watcher.go]: update comment.

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**